packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+    let corner = Corner.TOP_LEFT;
+
+    const {viewportDistance, anchorSize, surfaceSize} = this.measurements_;
+
+    const isBottomAligned = Boolean(this.anchorCorner_ & CornerBit.BOTTOM);
+    const availableTop = isBottomAligned ? viewportDistance.top + anchorSize.height + this.anchorMargin_.bottom
+      : viewportDistance.top + this.anchorMargin_.top;
+    const availableBottom = isBottomAligned ? viewportDistance.bottom - this.anchorMargin_.bottom
+      : viewportDistance.bottom + anchorSize.height - this.anchorMargin_.top;
+
+    const topOverflow = surfaceSize.height - availableTop;
+    const bottomOverflow = surfaceSize.height - availableBottom;
+    if (bottomOverflow > 0 && topOverflow < bottomOverflow) {
+      corner |= CornerBit.BOTTOM;
+    }
+
+    const isRtl = this.adapter_.isRtl();
+    const isFlipRtl = Boolean(this.anchorCorner_ & CornerBit.FLIP_RTL);
+    const avoidHorizontalOverlap = Boolean(this.anchorCorner_ & CornerBit.RIGHT);
+    const isAlignedRight = (avoidHorizontalOverlap && !isRtl) ||
+      (!avoidHorizontalOverlap && isFlipRtl && isRtl);
+    const availableLeft = isAlignedRight ? viewportDistance.left + anchorSize.width + this.anchorMargin_.right :
+      viewportDistance.left + this.anchorMargin_.left;
+    const availableRight = isAlignedRight ? viewportDistance.right - this.anchorMargin_.right :
+      viewportDistance.right + anchorSize.width - this.anchorMargin_.left;
+
+    const leftOverflow = surfaceSize.width - availableLeft;
+    const rightOverflow = surfaceSize.width - availableRight;
+
+    if ((leftOverflow < 0 && isAlignedRight && isRtl) ||
+        (avoidHorizontalOverlap && !isAlignedRight && leftOverflow < 0) ||
+        (rightOverflow > 0 && leftOverflow < rightOverflow)) {
+      corner |= CornerBit.RIGHT;
+    }
+
+    return corner;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
+   * @return Maximum height of the menu surface, based on available space. 0 indicates should not be set.
+   */
+  private getMenuSurfaceMaxHeight_(corner: Corner): number {
+    const {viewportDistance} = this.measurements_;
+
+    let maxHeight = 0;
+    const isBottomAligned = Boolean(corner & CornerBit.BOTTOM);
+    const {MARGIN_TO_EDGE} = MDCMenuSurfaceFoundation.numbers;
+
+    // When maximum height is not specified, it is handled from css.
+    if (isBottomAligned) {
+      maxHeight = viewportDistance.top + this.anchorMargin_.top - MARGIN_TO_EDGE;
+      if (!(this.anchorCorner_ & CornerBit.BOTTOM)) {
+        maxHeight += this.measurements_.anchorSize.height;
+      }
+    } else {
+      maxHeight =
+          viewportDistance.bottom - this.anchorMargin_.bottom + this.measurements_.anchorSize.height - MARGIN_TO_EDGE;
+      if (this.anchorCorner_ & CornerBit.BOTTOM) {
+        maxHeight -= this.measurements_.anchorSize.height;
+      }
+    }
+
+    return maxHeight;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
```

Comment:
remove closure comments

Reply:
This is actually [TypeDoc](https://typedoc.org/guides/doccomments/) syntax 😀